server/src/main/java/io/crate/statistics/ReservoirSampler.java

```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
     private CircuitBreakerService circuitBreakerService;
     private final IndicesService indicesService;
 
+    private volatile ByteSizeValue readLimit;
+    private volatile ByteSizeValue updatedReadLimit;
+    private volatile boolean analyzing;
+
+    private final RateLimiter rateLimiter;
+    private long bytesSinceLastRateLimit = 0;
+
     @Inject
     public ReservoirSampler(ClusterService clusterService,
                             NodeContext nodeCtx,
                             Schemas schemas,
                             CircuitBreakerService circuitBreakerService,
-                            IndicesService indicesService) {
+                            IndicesService indicesService,
+                            Settings settings) {
         this.clusterService = clusterService;
         this.nodeCtx = nodeCtx;
         this.schemas = schemas;
         this.circuitBreakerService = circuitBreakerService;
         this.indicesService = indicesService;
+
+        this.readLimit = STATS_SERVICE_THROTTLING_SETTING.get(settings);
+        this.rateLimiter = new RateLimiter.SimpleRateLimiter(readLimit.getMbFrac());
+        this.updatedReadLimit = readLimit;
+        this.analyzing = false;
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(
+            STATS_SERVICE_THROTTLING_SETTING, this::updateReadLimit);
+    }
+
+    private void updateReadLimit(ByteSizeValue newReadLimit) {
+        boolean isAnalyzing = analyzing; // One volatile read.
+        if (isAnalyzing == false) {
+            readLimit = newReadLimit;
+            rateLimiter.setMBPerSec(readLimit.getMbFrac());
+        } else {
+            updatedReadLimit = newReadLimit; // One volatile write. Let running ANALYZE operation finish with the previous value.
+        }
+    }
+
+    public void maybePause(long bytes) {
```

Comment:
this is derived from https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/index/snapshots/blobstore/RateLimitingInputStream.java#L49

can actually refactor/reuse if this approach is fine in general

Reply:
wouldn't touch `RateLimitingInoutStream` as I changed `maybePause` in https://github.com/crate/crate/commit/ef4e45b308401b4851e502b5a1d80b3b1d6d24d1 to accept current state in order to avoid counter synchronization.